tag_tracer.go

```diff
@@ -0,0 +1,300 @@
+	decaying map[string]connmgr.DecayingTag
+	direct   map[peer.ID]struct{}
+
+	// track message deliveries to reward "near first" deliveries
+	// (a delivery that occurs while we're still validating the message)
+	deliveries *messageDeliveries
+}
+
+func newTagTracer(cmgr connmgr.ConnManager) *tagTracer {
+	decayer, ok := connmgr.SupportsDecay(cmgr)
+	if !ok {
+		log.Warnf("connection manager does not support decaying tags, delivery tags will not be applied")
+	}
+	return &tagTracer{
+		cmgr:       cmgr,
+		msgID:      DefaultMsgIdFn,
+		decayer:    decayer,
+		decaying:   make(map[string]connmgr.DecayingTag),
+		deliveries: &messageDeliveries{records: make(map[string]*deliveryRecord)},
+	}
+}
+
+func (t *tagTracer) Start(gs *GossipSubRouter) {
+	if t == nil {
+		return
+	}
+
+	t.msgID = gs.p.msgID
+	t.direct = gs.direct
+	go t.background(gs.p.ctx)
+}
+
+func (t *tagTracer) background(ctx context.Context) {
+	gcDeliveryRecords := time.NewTicker(time.Minute)
+	defer gcDeliveryRecords.Stop()
+
+	for {
+		select {
+		case <-gcDeliveryRecords.C:
+			t.gcDeliveryRecords()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (t *tagTracer) gcDeliveryRecords() {
+	t.Lock()
+	defer t.Unlock()
+	t.deliveries.gc()
+}
+
+func (t *tagTracer) tagPeerIfDirect(p peer.ID) {
+	if t.direct == nil {
+		return
+	}
+
+	// tag peer if it is a direct peer
+	_, direct := t.direct[p]
+	if direct {
+		t.cmgr.TagPeer(p, "pubsub:direct", GossipSubConnTagValueDirectPeer)
+	}
+}
+
+func (t *tagTracer) tagMeshPeer(p peer.ID, topic string) {
+	tag := topicTag(topic)
+	t.cmgr.TagPeer(p, tag, GossipSubConnTagValueMeshPeer)
+}
+
+func (t *tagTracer) untagMeshPeer(p peer.ID, topic string) {
+	tag := topicTag(topic)
+	t.cmgr.UntagPeer(p, tag)
+}
+
+func topicTag(topic string) string {
+	return fmt.Sprintf("pubsub:%s", topic)
+}
+
+func (t *tagTracer) addDeliveryTag(topic string) {
+	if t.decayer == nil {
+		return
+	}
+
+	t.Lock()
+	defer t.Unlock()
+	tag, err := t.decayingDeliveryTag(topic)
+	if err != nil {
+		log.Warnf("unable to create decaying delivery tag: %s", err)
+		return
+	}
+	t.decaying[topic] = tag
+}
+
+func (t *tagTracer) removeDeliveryTag(topic string) {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.decaying, topic)
+}
+
+func (t *tagTracer) decayingDeliveryTag(topic string) (connmgr.DecayingTag, error) {
+	if t.decayer == nil {
+		return nil, fmt.Errorf("connection manager does not support decaying tags")
+	}
+	name := fmt.Sprintf("pubsub-deliveries:%s", topic)
+
+	// decrement tag value by GossipSubConnTagDecayAmount at each decay interval
+	decayFn := func(value connmgr.DecayingValue) (after int, rm bool) {
+		v := value.Value - GossipSubConnTagDecayAmount
+		return v, v <= 0
+	}
+
+	// bump up to max of GossipSubConnTagMessageDeliveryCap
+	bumpFn := func(value connmgr.DecayingValue, delta int) (after int) {
+		val := value.Value + delta
+		if val > GossipSubConnTagMessageDeliveryCap {
+			return GossipSubConnTagMessageDeliveryCap
+		}
+		return val
+	}
```

Comment:
I'm moving the [preset decay and bump functions](https://github.com/libp2p/go-libp2p-connmgr/pull/61/files#diff-7167ce1adea9dc2edd55d7253ca2d1f5) from `go-libp2p-connmgr` into `go-libp2p-core` so that you can reuse them here without introducing an additional module dependency.

What you've implemented here is basically the `FixedDecay` decay function, and the `SumBounded` bump function.

Reply:
Done. Mind adopting the preset functions here? https://github.com/libp2p/go-libp2p-core/pull/104/commits/d3277047d6ca1eae46b03c4816ca00e696b114e8